// @wordpress/theme/build-module/color-ramps/lib/constants.mjs
// packages/theme/src/color-ramps/lib/constants.ts
import { to, OKLCH } from "colorjs.io/fn";
import "./register-color-spaces.mjs";
var WHITE = to("white", OKLCH);
var BLACK = to("black", OKLCH);
var UNIVERSAL_CONTRAST_TOPUP = 0.02;
var WHITE_TEXT_CONTRAST_MARGIN = 3.1;
var ACCENT_SCALE_BASE_LIGHTNESS_THRESHOLDS = {
  lighter: { min: 0.2, max: 0.4 },
  darker: { min: 0.75, max: 0.98 }
};
var CONTRAST_EPSILON = 4e-3;
var MAX_BISECTION_ITERATIONS = 10;
var CONTRAST_COMBINATIONS = [
  {
    bgs: ["surface1", "surface2", "surface3"],
    fgs: ["fgSurface3", "fgSurface4"],
    target: 4.5
  },
  {
    bgs: ["surface4", "surface5"],
    fgs: ["fgSurface4"],
    target: 4.5
  },
  {
    bgs: ["bgFill1"],
    fgs: ["fgFill"],
    target: 4.5
  },
  {
    bgs: ["bgFillInverted1"],
    fgs: ["fgFillInverted"],
    target: 4.5
  },
  {
    bgs: ["bgFillInverted1"],
    fgs: ["fgFillInverted"],
    target: 4.5
  },
  {
    bgs: ["surface1", "surface2", "surface3"],
    fgs: ["stroke3"],
    target: 3
  }
];
var DEFAULT_SEED_COLORS = {
  bg: "#fcfcfc",
  primary: "#3858e9",
  info: "#0090ff",
  success: "#4ab866",
  caution: "#f0d149",
  warning: "#f0b849",
  error: "#cc1818"
};
export {
  ACCENT_SCALE_BASE_LIGHTNESS_THRESHOLDS,
  BLACK,
  CONTRAST_COMBINATIONS,
  CONTRAST_EPSILON,
  DEFAULT_SEED_COLORS,
  MAX_BISECTION_ITERATIONS,
  UNIVERSAL_CONTRAST_TOPUP,
  WHITE,
  WHITE_TEXT_CONTRAST_MARGIN
};
