// @wordpress/theme/build-module/color-ramps/lib/find-color-with-constraints.mjs
// packages/theme/src/color-ramps/lib/find-color-with-constraints.ts
import { get, OKLCH } from "colorjs.io/fn";
import "./register-color-spaces.mjs";
import { solveWithBisect } from "./utils.mjs";
import { WHITE, BLACK, CONTRAST_EPSILON } from "./constants.mjs";
import { clampToGamut, getContrast } from "./color-utils.mjs";
import { taperChroma } from "./taper-chroma.mjs";
function cdiff(c1, c2) {
  return Math.log(c1 / c2);
}
function findColorMeetingRequirements(reference, seed, target, direction, {
  lightnessConstraint,
  taperChromaOptions
} = {}) {
  if (target <= 1) {
    return {
      color: reference,
      reached: true,
      achieved: 1
    };
  }
  function getColorForL(l) {
    let newL = l;
    let newC = get(seed, [OKLCH, "c"]);
    if (taperChromaOptions) {
      const tapered = taperChroma(seed, newL, taperChromaOptions);
      if ("l" in tapered && "c" in tapered) {
        newL = tapered.l;
        newC = tapered.c;
      } else {
        return tapered;
      }
    }
    return clampToGamut({
      spaceId: "oklch",
      coords: [newL, newC, get(seed, [OKLCH, "h"])]
    });
  }
  const mostContrastingL = direction === "lighter" ? 1 : 0;
  const mostContrastingColor = direction === "lighter" ? WHITE : BLACK;
  const highestContrast = getContrast(reference, mostContrastingColor);
  if (lightnessConstraint) {
    const colorWithExactL = getColorForL(lightnessConstraint.value);
    const exactLContrast = getContrast(reference, colorWithExactL);
    const exactLContrastMeetsTarget = cdiff(exactLContrast, target) >= -CONTRAST_EPSILON;
    if (exactLContrastMeetsTarget || lightnessConstraint.type === "force") {
      return {
        color: colorWithExactL,
        reached: exactLContrastMeetsTarget,
        achieved: exactLContrast,
        deficit: exactLContrastMeetsTarget ? cdiff(exactLContrast, highestContrast) : cdiff(target, exactLContrast)
      };
    }
  }
  if (cdiff(highestContrast, target) <= CONTRAST_EPSILON) {
    return {
      color: mostContrastingColor,
      reached: cdiff(highestContrast, target) >= -CONTRAST_EPSILON,
      achieved: highestContrast,
      deficit: cdiff(target, highestContrast)
    };
  }
  const lowerL = get(reference, [OKLCH, "l"]);
  const lowerContrast = cdiff(1, target);
  const upperL = mostContrastingL;
  const upperContrast = cdiff(highestContrast, target);
  const bestColor = solveWithBisect(
    getColorForL,
    (c) => cdiff(getContrast(reference, c), target),
    lowerL,
    lowerContrast,
    upperL,
    upperContrast
  );
  return {
    color: bestColor,
    reached: true,
    achieved: target,
    // Negative number that specifies how much room we have.
    deficit: cdiff(target, highestContrast)
  };
}
export {
  findColorMeetingRequirements
};
