// @wordpress/theme/build-module/color-ramps/lib/utils.mjs
// packages/theme/src/color-ramps/lib/utils.ts
import "./register-color-spaces.mjs";
import {
  WHITE,
  BLACK,
  UNIVERSAL_CONTRAST_TOPUP,
  WHITE_TEXT_CONTRAST_MARGIN,
  ACCENT_SCALE_BASE_LIGHTNESS_THRESHOLDS,
  MAX_BISECTION_ITERATIONS,
  CONTRAST_EPSILON
} from "./constants.mjs";
import { getContrast } from "./color-utils.mjs";
function buildDependencyGraph(config) {
  const dependencies = /* @__PURE__ */ new Map();
  const dependents = /* @__PURE__ */ new Map();
  Object.keys(config).forEach((step) => {
    dependencies.set(step, []);
  });
  dependents.set("seed", []);
  Object.keys(config).forEach((step) => {
    dependents.set(step, []);
  });
  Object.entries(config).forEach(([stepName, stepConfig]) => {
    const step = stepName;
    const reference = stepConfig.contrast.reference;
    dependencies.get(step).push(reference);
    dependents.get(reference).push(step);
    if (stepConfig.sameAsIfPossible) {
      dependencies.get(step).push(stepConfig.sameAsIfPossible);
      dependents.get(stepConfig.sameAsIfPossible).push(step);
    }
  });
  return { dependencies, dependents };
}
function sortByDependency(config) {
  const { dependents } = buildDependencyGraph(config);
  const result = [];
  const visited = /* @__PURE__ */ new Set();
  const visiting = /* @__PURE__ */ new Set();
  function visit(node) {
    if (visiting.has(node)) {
      throw new Error(
        `Circular dependency detected involving step: ${String(
          node
        )}`
      );
    }
    if (visited.has(node)) {
      return;
    }
    visiting.add(node);
    const nodeDependents = dependents.get(node) || [];
    nodeDependents.forEach((dependent) => {
      visit(dependent);
    });
    visiting.delete(node);
    visited.add(node);
    if (node !== "seed") {
      result.unshift(node);
    }
  }
  visit("seed");
  return result;
}
function stepsForStep(stepName, config) {
  const result = /* @__PURE__ */ new Set();
  function visit(step) {
    if (step === "seed" || result.has(step)) {
      return;
    }
    const stepConfig = config[step];
    if (!stepConfig) {
      return;
    }
    visit(stepConfig.contrast.reference);
    if (stepConfig.sameAsIfPossible) {
      visit(stepConfig.sameAsIfPossible);
    }
    result.add(step);
  }
  visit(stepName);
  return Array.from(result);
}
function computeBetterFgColorDirection(seed, preferLighter) {
  const contrastAgainstBlack = getContrast(seed, BLACK);
  const contrastAgainstWhite = getContrast(seed, WHITE);
  return contrastAgainstBlack > contrastAgainstWhite + (preferLighter ? WHITE_TEXT_CONTRAST_MARGIN : 0) ? { better: "darker", worse: "lighter" } : { better: "lighter", worse: "darker" };
}
function adjustContrastTarget(target) {
  if (target === 1) {
    return 1;
  }
  return target + UNIVERSAL_CONTRAST_TOPUP;
}
function clampAccentScaleReferenceLightness(rawLightness, direction) {
  const thresholds = ACCENT_SCALE_BASE_LIGHTNESS_THRESHOLDS[direction];
  return Math.max(thresholds.min, Math.min(thresholds.max, rawLightness));
}
function solveWithBisect(calculateC, calculateValue, initLowerL, initLowerValue, initUpperL, initUpperValue) {
  let lowerL = initLowerL;
  let lowerValue = initLowerValue;
  let lowerReplaced = false;
  let upperL = initUpperL;
  let upperValue = initUpperValue;
  let upperReplaced = false;
  let bestC;
  let bestValue;
  let iterations = 0;
  while (true) {
    iterations++;
    const newL = (lowerL * upperValue - upperL * lowerValue) / (upperValue - lowerValue);
    bestC = calculateC(newL);
    bestValue = calculateValue(bestC);
    if (Math.abs(bestValue) <= CONTRAST_EPSILON || iterations >= MAX_BISECTION_ITERATIONS) {
      break;
    }
    if (bestValue <= 0) {
      lowerL = newL;
      lowerValue = bestValue;
      if (lowerReplaced) {
        upperValue /= 2;
      }
      lowerReplaced = true;
      upperReplaced = false;
    } else {
      upperL = newL;
      upperValue = bestValue;
      if (upperReplaced) {
        lowerValue /= 2;
      }
      upperReplaced = true;
      lowerReplaced = false;
    }
  }
  return bestC;
}
export {
  adjustContrastTarget,
  clampAccentScaleReferenceLightness,
  computeBetterFgColorDirection,
  solveWithBisect,
  sortByDependency,
  stepsForStep
};
