// @wordpress/theme/build-module/color-ramps/index.mjs
// packages/theme/src/color-ramps/index.ts
import { get, OKLCH } from "colorjs.io/fn";
import "./lib/register-color-spaces.mjs";
import { buildRamp } from "./lib/index.mjs";
import { clampAccentScaleReferenceLightness } from "./lib/utils.mjs";
import { BG_RAMP_CONFIG, ACCENT_RAMP_CONFIG } from "./lib/ramp-configs.mjs";
import { getContrast } from "./lib/color-utils.mjs";
import { CONTRAST_COMBINATIONS } from "./lib/constants.mjs";
import { DEFAULT_SEED_COLORS } from "./lib/constants.mjs";
function buildBgRamp(seed) {
  if (typeof seed !== "string" || seed.trim() === "") {
    throw new Error("Seed color must be a non-empty string");
  }
  return buildRamp(seed, BG_RAMP_CONFIG);
}
var STEP_TO_PIN = "surface2";
function getBgRampInfo(ramp) {
  return {
    mainDirection: ramp.direction,
    pinLightness: {
      stepName: STEP_TO_PIN,
      value: clampAccentScaleReferenceLightness(
        get(ramp.ramp[STEP_TO_PIN], [OKLCH, "l"]),
        ramp.direction
      )
    }
  };
}
function buildAccentRamp(seed, bgRamp) {
  if (typeof seed !== "string" || seed.trim() === "") {
    throw new Error("Seed color must be a non-empty string");
  }
  const bgRampInfo = bgRamp ? getBgRampInfo(bgRamp) : void 0;
  return buildRamp(seed, ACCENT_RAMP_CONFIG, bgRampInfo);
}
function checkAccessibleCombinations({
  bgRamp,
  accentRamps = []
}) {
  const unmetTargets = [];
  [bgRamp, ...accentRamps].forEach((ramp) => {
    CONTRAST_COMBINATIONS.forEach(({ bgs, fgs, target }) => {
      for (const bg of bgs) {
        for (const fg of fgs) {
          const bgColor = ramp.ramp[bg];
          const fgColor = ramp.ramp[fg];
          const achievedContrast = getContrast(bgColor, fgColor);
          if (achievedContrast < target) {
            unmetTargets.push({
              bgName: bg,
              bgColor,
              fgName: fg,
              fgColor,
              unmetContrast: target,
              achievedContrast
            });
          }
        }
      }
    });
  });
  accentRamps.forEach((ramp) => {
    CONTRAST_COMBINATIONS.forEach(({ bgs, fgs, target }) => {
      for (const bg of bgs) {
        for (const fg of fgs) {
          const bgColor = bgRamp.ramp[bg];
          const fgColor = ramp.ramp[fg];
          const achievedContrast = getContrast(bgColor, fgColor);
          if (achievedContrast < target) {
            unmetTargets.push({
              bgName: bg,
              bgColor,
              fgName: fg,
              fgColor,
              unmetContrast: target,
              achievedContrast
            });
          }
        }
      }
    });
  });
  return unmetTargets;
}
export {
  DEFAULT_SEED_COLORS,
  buildAccentRamp,
  buildBgRamp,
  checkAccessibleCombinations
};
