// @wordpress/e2e-test-utils-playwright/build/request-utils/media.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.listMedia = listMedia;
exports.uploadMedia = uploadMedia;
exports.deleteMedia = deleteMedia;
exports.deleteAllMedia = deleteAllMedia;
/**
 * External dependencies
 */
const fs = __importStar(require("fs"));
/**
 * List all media files.
 *
 * @see https://developer.wordpress.org/rest-api/reference/media/#list-media
 * @param this
 */
async function listMedia() {
    const response = await this.rest({
        method: 'GET',
        path: '/wp/v2/media',
        params: {
            per_page: 100,
        },
    });
    return response;
}
/**
 * Upload a media file.
 *
 * @see https://developer.wordpress.org/rest-api/reference/media/#create-a-media-item
 * @param this
 * @param filePathOrData The path or data of the file being uploaded.
 */
async function uploadMedia(filePathOrData) {
    const file = typeof filePathOrData === 'string'
        ? fs.createReadStream(filePathOrData)
        : filePathOrData;
    const response = await this.rest({
        method: 'POST',
        path: '/wp/v2/media',
        multipart: {
            file,
        },
    });
    return response;
}
/**
 * delete a media file.
 *
 * @see https://developer.wordpress.org/rest-api/reference/media/#delete-a-media-item
 * @param this
 * @param mediaId The ID of the media file.
 */
async function deleteMedia(mediaId) {
    const response = await this.rest({
        method: 'DELETE',
        path: `/wp/v2/media/${mediaId}`,
        params: { force: true },
    });
    return response;
}
/**
 * delete all media files.
 *
 * @param this
 */
async function deleteAllMedia() {
    const files = await this.listMedia();
    // The media endpoint doesn't support batch request yet.
    const responses = await Promise.all(files.map((media) => this.deleteMedia(media.id)));
    return responses;
}
