// @wordpress/theme/build/use-theme-provider-styles.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to2, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to2, key) && key !== except)
        __defProp(to2, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to2;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/theme/src/use-theme-provider-styles.ts
var use_theme_provider_styles_exports = {};
__export(use_theme_provider_styles_exports, {
  useThemeProviderStyles: () => useThemeProviderStyles
});
module.exports = __toCommonJS(use_theme_provider_styles_exports);
var import_fn = require("colorjs.io/fn");
var import_memize = __toESM(require("memize"));
var import_element = require("@wordpress/element");
var import_register_color_spaces = require("./color-ramps/lib/register-color-spaces.cjs");
var import_context = require("./context.cjs");
var import_color_tokens = __toESM(require("./prebuilt/ts/color-tokens.cjs"));
var import_color_ramps = require("./color-ramps/index.cjs");
var import_color_utils = require("./color-ramps/lib/color-utils.cjs");
var getCachedBgRamp = (0, import_memize.default)(import_color_ramps.buildBgRamp, { maxSize: 10 });
var getCachedAccentRamp = (0, import_memize.default)(import_color_ramps.buildAccentRamp, { maxSize: 10 });
var legacyWpComponentsOverridesCSS = [
  ["--wp-components-color-accent", "var(--wp-admin-theme-color)"],
  [
    "--wp-components-color-accent-darker-10",
    "var(--wp-admin-theme-color-darker-10)"
  ],
  [
    "--wp-components-color-accent-darker-20",
    "var(--wp-admin-theme-color-darker-20)"
  ],
  [
    "--wp-components-color-accent-inverted",
    "var(--wpds-color-fg-interactive-brand-strong, #fff)"
  ],
  [
    "--wp-components-color-background",
    "var(--wpds-color-bg-surface-neutral-strong, #ffffff)"
  ],
  [
    "--wp-components-color-foreground",
    "var(--wpds-color-fg-content-neutral, #1e1e1e)"
  ],
  [
    "--wp-components-color-foreground-inverted",
    "var(--wpds-color-bg-surface-neutral, #fcfcfc)"
  ],
  [
    "--wp-components-color-gray-100",
    "var(--wpds-color-bg-surface-neutral, #fcfcfc)"
  ],
  [
    "--wp-components-color-gray-200",
    "var(--wpds-color-stroke-surface-neutral, #dbdbdb)"
  ],
  [
    "--wp-components-color-gray-300",
    "var(--wpds-color-stroke-surface-neutral, #dbdbdb)"
  ],
  [
    "--wp-components-color-gray-400",
    "var(--wpds-color-stroke-interactive-neutral, #8d8d8d)"
  ],
  [
    "--wp-components-color-gray-600",
    "var(--wpds-color-stroke-interactive-neutral, #8d8d8d)"
  ],
  [
    "--wp-components-color-gray-700",
    "var(--wpds-color-fg-content-neutral-weak, #707070)"
  ],
  [
    "--wp-components-color-gray-800",
    "var(--wpds-color-fg-content-neutral, #1e1e1e)"
  ]
];
function customRgbFormat(color) {
  const rgb = (0, import_fn.to)(color, import_fn.sRGB);
  return rgb.coords.map((n) => Math.round((n ?? 0) * 255)).join(", ");
}
function legacyWpAdminThemeOverridesCSS(accent) {
  const parsedAccent = (0, import_fn.to)(accent, import_fn.HSL);
  const parsedL = parsedAccent.coords[2] ?? 0;
  const darker10 = (0, import_fn.set)(
    (0, import_fn.clone)(parsedAccent),
    [import_fn.HSL, "l"],
    Math.max(0, parsedL - 5)
    // L reduced by 5%
  );
  const darker20 = (0, import_fn.set)(
    (0, import_fn.clone)(parsedAccent),
    [import_fn.HSL, "l"],
    Math.max(0, parsedL - 10)
    // L reduced by 10%
  );
  return [
    ["--wp-admin-theme-color", (0, import_color_utils.getColorString)(parsedAccent)],
    ["--wp-admin-theme-color--rgb", customRgbFormat(parsedAccent)],
    ["--wp-admin-theme-color-darker-10", (0, import_color_utils.getColorString)(darker10)],
    [
      "--wp-admin-theme-color-darker-10--rgb",
      customRgbFormat(darker10)
    ],
    ["--wp-admin-theme-color-darker-20", (0, import_color_utils.getColorString)(darker20)],
    [
      "--wp-admin-theme-color-darker-20--rgb",
      customRgbFormat(darker20)
    ]
  ];
}
function colorTokensCSS(computedColorRamps) {
  const entries = [];
  for (const [rampName, { ramp }] of computedColorRamps) {
    for (const [tokenName, tokenValue] of Object.entries(ramp)) {
      const key = `${rampName}-${tokenName}`;
      const aliasedBy = import_color_tokens.default[key] ?? [];
      for (const aliasedId of aliasedBy) {
        entries.push([`--wpds-color-${aliasedId}`, tokenValue]);
      }
    }
  }
  return entries;
}
function generateStyles({
  primary,
  computedColorRamps
}) {
  return Object.fromEntries(
    [
      // Semantic color tokens
      colorTokensCSS(computedColorRamps),
      // Legacy overrides
      legacyWpAdminThemeOverridesCSS(primary),
      legacyWpComponentsOverridesCSS
    ].flat()
  );
}
function useThemeProviderStyles({
  color = {},
  cursor
} = {}) {
  const { resolvedSettings: inheritedSettings } = (0, import_element.useContext)(import_context.ThemeContext);
  const primary = color.primary ?? inheritedSettings.color?.primary ?? import_color_ramps.DEFAULT_SEED_COLORS.primary;
  const bg = color.bg ?? inheritedSettings.color?.bg ?? import_color_ramps.DEFAULT_SEED_COLORS.bg;
  const cursorControl = cursor?.control ?? inheritedSettings.cursor?.control;
  const resolvedSettings = (0, import_element.useMemo)(
    () => ({
      color: {
        primary,
        bg
      },
      cursor: cursorControl ? { control: cursorControl } : void 0
    }),
    [primary, bg, cursorControl]
  );
  const colorStyles = (0, import_element.useMemo)(() => {
    const seeds = {
      ...import_color_ramps.DEFAULT_SEED_COLORS,
      bg,
      primary
    };
    const computedColorRamps = /* @__PURE__ */ new Map();
    const bgRamp = getCachedBgRamp(seeds.bg);
    Object.entries(seeds).forEach(([rampName, seed]) => {
      if (rampName === "bg") {
        computedColorRamps.set(rampName, bgRamp);
      } else {
        computedColorRamps.set(
          rampName,
          getCachedAccentRamp(seed, bgRamp)
        );
      }
    });
    return generateStyles({
      primary: seeds.primary,
      computedColorRamps
    });
  }, [primary, bg]);
  const themeProviderStyles = (0, import_element.useMemo)(
    () => ({
      ...colorStyles,
      ...cursorControl && {
        "--wpds-cursor-control": cursorControl
      }
    }),
    [colorStyles, cursorControl]
  );
  return {
    resolvedSettings,
    themeProviderStyles
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  useThemeProviderStyles
});
