// @wordpress/theme/src/color-ramps/lib/default-ramps.ts
/**
 * Internal dependencies
 */
import type { RampResult } from './types';
import type { DEFAULT_SEED_COLORS } from './constants';

export const DEFAULT_RAMPS: Record<
	keyof typeof DEFAULT_SEED_COLORS,
	RampResult
> = {
	bg: {
		ramp: {
			surface2: '#fcfcfc',
			bgFill1: '#555',
			fgFill: '#f0f0f0',
			bgFill2: '#484848',
			surface6: '#d6d6d6',
			surface5: '#e6e6e6',
			surface4: '#ededed',
			surface3: '#fff',
			fgSurface4: '#1e1e1e',
			fgSurface3: '#707070',
			fgSurface2: '#8d8d8d',
			fgSurface1: '#aeaeae',
			stroke3: '#8d8d8d',
			stroke4: '#6e6e6e',
			stroke2: '#dbdbdb',
			stroke1: '#e4e4e4',
			bgFillDark: '#1e1e1e',
			fgFillDark: '#f0f0f0',
			bgFillInverted2: '#1e1e1e',
			bgFillInverted1: '#2d2d2d',
			fgFillInverted: '#f0f0f0',
			surface1: '#f4f4f4',
		},
		direction: 'darker',
	},
	primary: {
		ramp: {
			bgFill1: '#3858e9',
			fgFill: '#eff0f2',
			bgFill2: '#2e49d9',
			surface2: '#f6f8fe',
			surface6: '#c7d2ee',
			surface5: '#dce3f4',
			surface4: '#e6eaf4',
			surface3: '#fff',
			fgSurface4: '#0b0070',
			fgSurface3: '#3858e9',
			fgSurface2: '#5a82ff',
			fgSurface1: '#86a9ff',
			stroke3: '#3858e9',
			stroke4: '#2337c8',
			stroke2: '#9caacc',
			stroke1: '#a3b1d4',
			bgFillDark: '#1a1e27',
			fgFillDark: '#eff0f2',
			bgFillInverted2: '#1a1e27',
			bgFillInverted1: '#13009f',
			fgFillInverted: '#eff0f2',
			surface1: '#ecf0fa',
		},
		direction: 'darker',
	},
	info: {
		ramp: {
			bgFill1: '#0090ff',
			fgFill: '#1a1f24',
			bgFill2: '#007fed',
			surface2: '#f3f9ff',
			surface6: '#bcd5f2',
			surface5: '#d4e4f7',
			surface4: '#deebfa',
			surface3: '#fff',
			fgSurface4: '#001b4f',
			fgSurface3: '#006bd7',
			fgSurface2: '#008bfa',
			fgSurface1: '#59b0ff',
			stroke3: '#006bd7',
			stroke4: '#004fa9',
			stroke2: '#95b5d9',
			stroke1: '#9fbcdc',
			bgFillDark: '#1a1f24',
			fgFillDark: '#eef0f3',
			bgFillInverted2: '#1a1f24',
			bgFillInverted1: '#002a69',
			fgFillInverted: '#eef0f3',
			surface1: '#e5f1ff',
		},
		direction: 'darker',
	},
	success: {
		ramp: {
			bgFill1: '#4ab866',
			fgFill: '#1b1f1c',
			bgFill2: '#37a756',
			surface2: '#ebffed',
			surface6: '#88e89b',
			surface5: '#b0f3bb',
			surface4: '#c6f7cd',
			surface3: '#fff',
			fgSurface4: '#002900',
			fgSurface3: '#008030',
			fgSurface2: '#2b9f4e',
			fgSurface1: '#54c16e',
			stroke3: '#008030',
			stroke4: '#006113',
			stroke2: '#84c08f',
			stroke1: '#8ac894',
			bgFillDark: '#1b1f1c',
			fgFillDark: '#eaf3eb',
			bgFillInverted2: '#1b1f1c',
			bgFillInverted1: '#003701',
			fgFillInverted: '#eaf3eb',
			surface1: '#ccfdd3',
		},
		direction: 'darker',
	},
	caution: {
		ramp: {
			bgFill1: '#f0d149',
			fgFill: '#1f1e1b',
			bgFill2: '#dcbe2f',
			surface2: '#fff9ca',
			surface6: '#e9d170',
			surface5: '#f7e28a',
			surface4: '#fee995',
			surface3: '#fff',
			fgSurface4: '#281d00',
			fgSurface3: '#826a00',
			fgSurface2: '#a48700',
			fgSurface1: '#c6a800',
			stroke3: '#826a00',
			stroke4: '#634f00',
			stroke2: '#bdb17e',
			stroke1: '#c5b883',
			bgFillDark: '#1f1e1b',
			fgFillDark: '#fdf1bf',
			bgFillInverted2: '#1f1e1b',
			bgFillInverted1: '#392c00',
			fgFillInverted: '#fdf1bf',
			surface1: '#fff09c',
		},
		direction: 'darker',
	},
	warning: {
		ramp: {
			bgFill1: '#f0b849',
			fgFill: '#1f1e1b',
			bgFill2: '#dda633',
			surface2: '#fff7e1',
			surface6: '#f4cc85',
			surface5: '#fadead',
			surface4: '#fde6be',
			surface3: '#fff',
			fgSurface4: '#2e1900',
			fgSurface3: '#926300',
			fgSurface2: '#b57f00',
			fgSurface1: '#d7a02b',
			stroke3: '#926300',
			stroke4: '#6f4900',
			stroke2: '#c7ad7e',
			stroke1: '#d0b481',
			bgFillDark: '#1f1e1b',
			fgFillDark: '#f7efe2',
			bgFillInverted2: '#1f1e1b',
			bgFillInverted1: '#422800',
			fgFillInverted: '#f7efe2',
			surface1: '#ffedc5',
		},
		direction: 'darker',
	},
	error: {
		ramp: {
			bgFill1: '#cc1818',
			fgFill: '#f2efef',
			bgFill2: '#b90000',
			surface2: '#fff6f5',
			surface6: '#f3c9c2',
			surface5: '#f7dbd7',
			surface4: '#f6e6e3',
			surface3: '#fff',
			fgSurface4: '#470000',
			fgSurface3: '#cc1818',
			fgSurface2: '#f74c40',
			fgSurface1: '#ff8879',
			stroke3: '#cc1818',
			stroke4: '#9d0000',
			stroke2: '#d39c95',
			stroke1: '#daa39b',
			bgFillDark: '#231c1b',
			fgFillDark: '#f2efef',
			bgFillInverted2: '#231c1b',
			bgFillInverted1: '#640000',
			fgFillInverted: '#f2efef',
			surface1: '#fcece9',
		},
		direction: 'darker',
	},
};
