// @wordpress/theme/build/stylelint-plugins/no-token-fallback-values.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/theme/src/stylelint-plugins/no-token-fallback-values.mjs
var no_token_fallback_values_exports = {};
__export(no_token_fallback_values_exports, {
  default: () => no_token_fallback_values_default
});
module.exports = __toCommonJS(no_token_fallback_values_exports);
var import_stylelint = __toESM(require("stylelint"), 1);
var {
  createPlugin,
  utils: { report, ruleMessages, validateOptions }
} = import_stylelint.default;
var ruleName = "plugin-wpds/no-token-fallback-values";
var varWithFallbackRegex = /var\(\s*(--wpds-[\w-]+)\s*,/g;
var messages = ruleMessages(ruleName, {
  rejected: (tokenName) => `Do not add a fallback value for Design System token '${tokenName}'. Fallbacks are injected automatically at build time.`
});
var ruleFunction = (primary) => {
  return (root, result) => {
    const validOptions = validateOptions(result, ruleName, {
      actual: primary,
      possible: [true]
    });
    if (!validOptions) {
      return;
    }
    root.walkDecls((ruleNode) => {
      const { value } = ruleNode;
      let match;
      varWithFallbackRegex.lastIndex = 0;
      while ((match = varWithFallbackRegex.exec(value)) !== null) {
        report({
          message: messages.rejected(match[1]),
          node: ruleNode,
          word: match[0],
          result,
          ruleName
        });
      }
    });
  };
};
ruleFunction.ruleName = ruleName;
ruleFunction.messages = messages;
var no_token_fallback_values_default = createPlugin(ruleName, ruleFunction);
