// @wordpress/theme/build/color-ramps/lib/ramp-configs.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/theme/src/color-ramps/lib/ramp-configs.ts
var ramp_configs_exports = {};
__export(ramp_configs_exports, {
  ACCENT_RAMP_CONFIG: () => ACCENT_RAMP_CONFIG,
  BG_RAMP_CONFIG: () => BG_RAMP_CONFIG
});
module.exports = __toCommonJS(ramp_configs_exports);
var lightnessConstraintForegroundHighContrast = (direction) => direction === "lighter" ? 0.9551 : 0.235;
var lightnessConstraintForegroundMediumContrast = (direction) => direction === "lighter" ? 0.77 : 0.56;
var lightnessConstraintBgFill = (direction) => direction === "lighter" ? 0.67 : 0.45;
var BG_SURFACE_TAPER_CHROMA = {
  alpha: 0.7
};
var FG_TAPER_CHROMA = {
  alpha: 0.6,
  kLight: 0.2,
  kDark: 0.2
};
var STROKE_TAPER_CHROMA = {
  alpha: 0.6,
  radiusDark: 0.01,
  radiusLight: 0.01,
  kLight: 0.8,
  kDark: 0.8
};
var ACCENT_SURFACE_TAPER_CHROMA = {
  alpha: 0.75,
  radiusDark: 0.01,
  radiusLight: 0.01
};
var fgSurface4Config = {
  contrast: {
    reference: "surface3",
    followDirection: "main",
    target: 7,
    preferLighter: true
  },
  lightness: lightnessConstraintForegroundHighContrast,
  taperChromaOptions: FG_TAPER_CHROMA
};
var BG_RAMP_CONFIG = {
  // Surface
  surface1: {
    contrast: {
      reference: "surface2",
      followDirection: "opposite",
      target: 1.06,
      ignoreWhenAdjustingSeed: true
    },
    taperChromaOptions: BG_SURFACE_TAPER_CHROMA
  },
  surface2: {
    contrast: {
      reference: "seed",
      followDirection: "main",
      target: 1
    }
  },
  surface3: {
    contrast: {
      reference: "surface2",
      followDirection: "main",
      target: 1.06
    },
    taperChromaOptions: BG_SURFACE_TAPER_CHROMA
  },
  surface4: {
    contrast: {
      reference: "surface2",
      followDirection: "main",
      target: 1.12
    },
    taperChromaOptions: BG_SURFACE_TAPER_CHROMA
  },
  surface5: {
    contrast: {
      reference: "surface2",
      followDirection: "main",
      target: 1.2
    },
    taperChromaOptions: BG_SURFACE_TAPER_CHROMA
  },
  surface6: {
    contrast: {
      reference: "surface2",
      followDirection: "main",
      target: 1.4
    },
    taperChromaOptions: BG_SURFACE_TAPER_CHROMA
  },
  // Bg fill
  bgFill1: {
    contrast: {
      reference: "surface2",
      followDirection: "main",
      target: 4
    },
    lightness: lightnessConstraintBgFill
  },
  bgFill2: {
    contrast: {
      reference: "bgFill1",
      followDirection: "main",
      target: 1.2
    }
  },
  bgFillInverted1: {
    contrast: {
      reference: "bgFillInverted2",
      followDirection: "opposite",
      target: 1.2
    }
  },
  bgFillInverted2: fgSurface4Config,
  bgFillDark: {
    contrast: {
      reference: "surface3",
      followDirection: "darker",
      // This is what causes the token to be always dark
      target: 7,
      ignoreWhenAdjustingSeed: true
    },
    lightness: lightnessConstraintForegroundHighContrast,
    taperChromaOptions: FG_TAPER_CHROMA
  },
  // Stroke
  stroke1: {
    contrast: {
      reference: "stroke3",
      followDirection: "opposite",
      target: 2.6
    },
    taperChromaOptions: STROKE_TAPER_CHROMA
  },
  stroke2: {
    contrast: {
      reference: "stroke3",
      followDirection: "opposite",
      target: 2.4
    },
    taperChromaOptions: STROKE_TAPER_CHROMA
  },
  stroke3: {
    contrast: {
      reference: "surface3",
      followDirection: "main",
      target: 3
    },
    taperChromaOptions: STROKE_TAPER_CHROMA
  },
  stroke4: {
    contrast: {
      reference: "stroke3",
      followDirection: "main",
      target: 1.5
    },
    taperChromaOptions: STROKE_TAPER_CHROMA
  },
  // fgSurface
  fgSurface1: {
    contrast: {
      reference: "surface3",
      followDirection: "main",
      target: 2,
      preferLighter: true
    },
    taperChromaOptions: FG_TAPER_CHROMA
  },
  fgSurface2: {
    contrast: {
      reference: "surface3",
      followDirection: "main",
      target: 3,
      preferLighter: true
    },
    taperChromaOptions: FG_TAPER_CHROMA
  },
  fgSurface3: {
    contrast: {
      reference: "surface3",
      followDirection: "main",
      target: 4.5,
      preferLighter: true
    },
    lightness: lightnessConstraintForegroundMediumContrast,
    taperChromaOptions: FG_TAPER_CHROMA
  },
  fgSurface4: fgSurface4Config,
  // fgFill
  fgFill: {
    contrast: {
      reference: "bgFill1",
      followDirection: "best",
      target: 4.5,
      preferLighter: true
    },
    lightness: lightnessConstraintForegroundHighContrast,
    taperChromaOptions: FG_TAPER_CHROMA
  },
  fgFillInverted: {
    contrast: {
      reference: "bgFillInverted1",
      followDirection: "best",
      target: 4.5,
      preferLighter: true
    },
    lightness: lightnessConstraintForegroundHighContrast,
    taperChromaOptions: FG_TAPER_CHROMA
  },
  fgFillDark: {
    contrast: {
      reference: "bgFillDark",
      followDirection: "best",
      target: 4.5,
      preferLighter: true
    },
    lightness: lightnessConstraintForegroundHighContrast,
    taperChromaOptions: FG_TAPER_CHROMA
  }
};
var ACCENT_RAMP_CONFIG = {
  ...BG_RAMP_CONFIG,
  surface1: {
    ...BG_RAMP_CONFIG.surface1,
    taperChromaOptions: ACCENT_SURFACE_TAPER_CHROMA
  },
  surface2: {
    contrast: {
      reference: "bgFill1",
      followDirection: "opposite",
      target: BG_RAMP_CONFIG.bgFill1.contrast.target,
      ignoreWhenAdjustingSeed: true
    },
    taperChromaOptions: ACCENT_SURFACE_TAPER_CHROMA
  },
  surface3: {
    ...BG_RAMP_CONFIG.surface3,
    taperChromaOptions: ACCENT_SURFACE_TAPER_CHROMA
  },
  surface4: {
    ...BG_RAMP_CONFIG.surface4,
    taperChromaOptions: ACCENT_SURFACE_TAPER_CHROMA
  },
  surface5: {
    ...BG_RAMP_CONFIG.surface5,
    taperChromaOptions: ACCENT_SURFACE_TAPER_CHROMA
  },
  surface6: {
    ...BG_RAMP_CONFIG.surface6,
    taperChromaOptions: ACCENT_SURFACE_TAPER_CHROMA
  },
  bgFill1: {
    contrast: {
      reference: "seed",
      followDirection: "main",
      target: 1
    }
  },
  stroke1: {
    ...BG_RAMP_CONFIG.stroke1
  },
  stroke2: {
    ...BG_RAMP_CONFIG.stroke2
  },
  stroke3: {
    ...BG_RAMP_CONFIG.stroke3,
    sameAsIfPossible: "fgSurface3",
    taperChromaOptions: void 0
  },
  stroke4: {
    ...BG_RAMP_CONFIG.stroke4,
    taperChromaOptions: void 0
  },
  // fgSurface: do not de-saturate
  fgSurface1: {
    ...BG_RAMP_CONFIG.fgSurface1,
    taperChromaOptions: void 0
  },
  fgSurface2: {
    ...BG_RAMP_CONFIG.fgSurface2,
    taperChromaOptions: void 0
  },
  fgSurface3: {
    ...BG_RAMP_CONFIG.fgSurface3,
    taperChromaOptions: void 0,
    sameAsIfPossible: "bgFill1"
  },
  fgSurface4: {
    ...BG_RAMP_CONFIG.fgSurface4,
    taperChromaOptions: void 0
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ACCENT_RAMP_CONFIG,
  BG_RAMP_CONFIG
});
