// @wordpress/theme/build-module/stylelint-plugins/no-token-fallback-values.mjs
// packages/theme/src/stylelint-plugins/no-token-fallback-values.mjs
import stylelint from "stylelint";
var {
  createPlugin,
  utils: { report, ruleMessages, validateOptions }
} = stylelint;
var ruleName = "plugin-wpds/no-token-fallback-values";
var varWithFallbackRegex = /var\(\s*(--wpds-[\w-]+)\s*,/g;
var messages = ruleMessages(ruleName, {
  rejected: (tokenName) => `Do not add a fallback value for Design System token '${tokenName}'. Fallbacks are injected automatically at build time.`
});
var ruleFunction = (primary) => {
  return (root, result) => {
    const validOptions = validateOptions(result, ruleName, {
      actual: primary,
      possible: [true]
    });
    if (!validOptions) {
      return;
    }
    root.walkDecls((ruleNode) => {
      const { value } = ruleNode;
      let match;
      varWithFallbackRegex.lastIndex = 0;
      while ((match = varWithFallbackRegex.exec(value)) !== null) {
        report({
          message: messages.rejected(match[1]),
          node: ruleNode,
          word: match[0],
          result,
          ruleName
        });
      }
    });
  };
};
ruleFunction.ruleName = ruleName;
ruleFunction.messages = messages;
var no_token_fallback_values_default = createPlugin(ruleName, ruleFunction);
export {
  no_token_fallback_values_default as default
};
